### PR TITLE
[SE-4860] make SINGLE_BEAT_IDENTIFIER configurable

### PR DESCRIPTION
**Pull request by gabor-boros**
## Description

This PR allows setting separate users for single-beat and celery brokers, allows authentication, and setting the single-beat identifier for future fine-grained access control using Redis ACLs.

Also, we are replacing the source url of the package, hence the original code became outdated and `akachanov:master` is ahead of the original repo by fixing issues and removing incompatible code.

## Supporting information

Some months ago edX platform was not capable of doing periodic scheduling of celery tasks. Therefore we added celerybeat to the platform. The only issue with celerybeat is that it cannot run more than once. If multiple celerybeat schedulers are running at the same time, the same task can be scheduled multiple times resulting in unwanted behaviours. Here comes `single-beat` into picture that ensures (using redis) that no more than one celerybeat processes are running.

> Can you explain what it is in akachanov's fork that's critical to us here?

Akachanov [removed](https://github.com/ybrs/single-beat/compare/master...akachanov:master#diff-fe06202dfa089ced8f935d1530b76acf71876400c17783f10a5ecf8c7ac3e67aL13) Tornadis that was published in 2018 for the last time and broke with the newest Celery. Using Akachanov's fork ensures that single-beat is still working with the newest Celery and he also seems to fixed some reconnect issues -- I assume hence the celerybeat process restarts, since I cannot see any errors and periodically scheduled tasks are processed as expected.

### Dependencies

Related to https://github.com/edx/edx-platform/pull/29046, which upgrades celery in edx-platform.

## Sandbox

[https://pr29046.sandbox.opencraft.hosting](https://pr29046.sandbox.opencraft.hosting/)

## Testing instructions

* Navigate to https://pr29046.sandbox.opencraft.hosting/heartbeat?extended and validate it is working properly
* Go to https://pr29046.sandbox.opencraft.hosting/admin/instructor_task/instructortask/ and validate reports are generated every 10 minutes

## Deadline

ASAP

## Other information

N/A

### CODE REVIEW

**Comment by openedx-webhooks** [comment]
Thanks for the pull request, @gabor-boros! I've created [OSPR-6145](https://openedx.atlassian.net/browse/OSPR-6145) to keep track of it in JIRA, where we prioritize reviews. Please note that it may take us up to several weeks or months to complete a review and merge your PR.

Feel free to add as much of the following information to the ticket as you can:
- supporting documentation
- [Open edX discussion forum threads](https://discuss.openedx.org/)
- timeline information ("this must be merged by XX date", and why that is)
- partner information ("this is a course on edx.org")
- any other information that can help Product understand the context for the PR

All technical communication about the code itself will be done via the GitHub pull request interface. As a reminder, [our process documentation is here](http://edx-developer-guide.readthedocs.org/en/latest/process/overview.html).


Please let us know once your PR is ready for our review and all tests are green.

**Comment by natabene** [comment]
@gabor-boros Thank you for your contribution. Please let me know once this is ready for our review.

File: playbooks/roles/edxapp/defaults/main.yml [hunk 1]
```diff
@@ -185,7 +185,11 @@ EDXAPP_ENABLE_CELERY_BEAT: false
 EDXAPP_SINGLE_BEAT_LOCK_TIME: 60
 # EDXAPP_SINGLE_BEAT_HEARTBEAT_INTERVAL must be smaller than EDXAPP_SINGLE_BEAT_LOCK_TIME / 2
 EDXAPP_SINGLE_BEAT_HEARTBEAT_INTERVAL: 29
-EDXAPP_SINGLE_BEAT_VERSION: "0.4.2"
+EDXAPP_SINGLE_BEAT_REPO: "https://github.com/akachanov/single-beat.git"
+EDXAPP_SINGLE_BEAT_VERSION: "e500ac4b56756cdf96836666883af8060aaef455"
```
Comment (pomegranited) [inline playbooks/roles/edxapp/defaults/main.yml:189]: @gabor-boros  I think it's going to be a hard sell to make this fork upstream's default source for `single-beat`, and unfortunately `akachanov` hasn't submitted his changes upstream yet either.

1. Can you explain what it is in `akachanov`'s fork that's critical to us here? [compare ybrs:master...akachanov:master](https://github.com/ybrs/single-beat/compare/master...akachanov:master)
2. Are the changes you've made to the single-beat invocation because of the changes in `akachanov`'s fork, or are they due to the celery upgrade?

Basically I want to know, can we retain `single-beat==0.4.2` and still work with single-tenant redis installs like edX's?  If so, then better to change the above to this:

```
EDXAPP_SINGLE_BEAT_REPO: https://github.com/ybrs/single-beat.git
EDXAPP_SINGLE_BEAT_VERSION: "66c8c6b9d5f405399654bcaa8f6dc0ae21bc404f"  # version 0.4.2
```

And use these overrides in our own code.

File: playbooks/roles/edx_django_service/templates/edx/app/supervisor/conf.d.available/app-workers.conf.j2 [hunk 2]
```diff
@@ -10,7 +10,7 @@ directory={{ edx_django_service_code_dir }}
 stdout_logfile={{ supervisor_log_dir }}/%(program_name)s-stdout.log
 stderr_logfile={{ supervisor_log_dir }}/%(program_name)s-stderr.log
 
-command={{ edx_django_service_home }}/{{ edx_django_service_name }}-workers.sh worker -A {{ edx_django_service_name }} --app {{ edx_django_service_name }}.celery:app --loglevel=info --queue={{ w.queue }} --hostname={{ edx_django_service_name }}.{{ w.queue }}.%%h --concurrency=1 {{ '--maxtasksperchild ' + w.max_tasks_per_child|string if w.max_tasks_per_child is defined else '' }} {{ '--without-heartbeat' if not edx_django_service_celery_heartbeat_enabled|bool else '' }}
+command={{ edx_django_service_home }}/{{ edx_django_service_name }}-workers.sh --app {{ edx_django_service_name }}.celery:app worker -A {{ edx_django_service_name }} --loglevel=info --queue={{ w.queue }} --hostname={{ edx_django_service_name }}.{{ w.queue }}.%%h --concurrency=1 {{ '--max-tasks-per-child ' + w.max_tasks_per_child|string if w.max_tasks_per_child is defined else '' }} {{ '--without-heartbeat' if not edx_django_service_celery_heartbeat_enabled|bool else '' }}
```
Comment (pomegranited) [inline playbooks/roles/edx_django_service/templates/edx/app/supervisor/conf.d.available/app-workers.conf.j2:13]: nit: Is it possible to reduce the diff here for clarity? Unless the order of arguments really matters, the only real diff I see here is changing `--maxtasksperchild` to `--max-tasks-per-child`?

```suggestion
command={{ edx_django_service_home }}/{{ edx_django_service_name }}-workers.sh worker -A {{ edx_django_service_name }} --app {{ edx_django_service_name }}.celery:app --loglevel=info --queue={{ w.queue }} --hostname={{ edx_django_service_name }}.{{ w.queue }}.%%h --concurrency=1 {{ '--max-tasks-per-child ' + w.max_tasks_per_child|string if w.max_tasks_per_child is defined else '' }} {{ '--without-heartbeat' if not edx_django_service_celery_heartbeat_enabled|bool else '' }}
```

File: playbooks/roles/edx_django_service/templates/edx/app/supervisor/conf.d.available/app-workers.conf.j2 [hunk 3]
```diff
@@ -39,7 +39,7 @@ stderr_logfile={{ supervisor_log_dir }}/%(program_name)s-stderr.log
 {% set worker_app_name = 'course_discovery' %}
 {% endif %}
 
-command={{ edx_django_service_home }}/{{ edx_django_service_name }}-workers.sh worker -A {{ edx_django_service_name }} --app {{ worker_app_name }}.celery:app --loglevel=info --queue={{ w.queue }} --hostname={{ edx_django_service_name }}.{{ w.queue }}.%%h --concurrency={{ w.concurrency }} {{ '--maxtasksperchild ' + w.max_tasks_per_child|string if w.max_tasks_per_child is defined else '' }} {{ '--without-heartbeat' if not edx_django_service_celery_heartbeat_enabled|bool else '' }}
+command={{ edx_django_service_home }}/{{ edx_django_service_name }}-workers.sh --app {{ worker_app_name }}.celery:app worker -A {{ edx_django_service_name }} --loglevel=info --queue={{ w.queue }} --hostname={{ edx_django_service_name }}.{{ w.queue }}.%%h --concurrency={{ w.concurrency }} {{ '--max-tasks-per-child ' + w.max_tasks_per_child|string if w.max_tasks_per_child is defined else '' }} {{ '--without-heartbeat' if not edx_django_service_celery_heartbeat_enabled|bool else '' }}
```
Comment (pomegranited) [inline playbooks/roles/edx_django_service/templates/edx/app/supervisor/conf.d.available/app-workers.conf.j2:37]: nit: same as [above](https://github.com/edx/configuration/pull/6586/files#diff-5ba60704de514ef153e8a1f6e9f86bd3fbdab429b4531ff135e3c02e2370030aR13).

File: playbooks/roles/edxapp/defaults/main.yml (hunk 1 above)
Comment (pomegranited) [reply playbooks/roles/edxapp/defaults/main.yml:189]: Also, the uptime for the `celerybeat_scheduler` on the sandbox isn't good.. is something wrong here?

```bash
pomegranited@edxapp-pr29046sandboxopencraft-appserver-22 i-01aff47e:~$ /edx/bin/supervisorctl status
celerybeat_scheduler             RUNNING   pid 27524, uptime 0:00:23
cms                              RUNNING   pid 940, uptime 14:17:32
edxapp_worker:cms_default_1      RUNNING   pid 944, uptime 14:17:32
edxapp_worker:cms_high_1         RUNNING   pid 945, uptime 14:17:32
edxapp_worker:lms_default_1      RUNNING   pid 953, uptime 14:17:32
edxapp_worker:lms_high_1         RUNNING   pid 960, uptime 14:17:31
edxapp_worker:lms_high_mem_1     RUNNING   pid 969, uptime 14:17:31
forum                            RUNNING   pid 1036, uptime 14:17:31
lms                              RUNNING   pid 1044, uptime 14:17:30
xqueue                           RUNNING   pid 1063, uptime 14:17:30
xqueue_consumer                  RUNNING   pid 1067, uptime 14:17:30
```

I can see that there are celerybeat tasks scheduled in the database [via the Django Admin](https://pr29046.sandbox.opencraft.hosting/admin/django_celery_beat/periodictask/), but what logs can we check to ensure they're actually running?

File: playbooks/roles/ecomworker/templates/edx/app/ecomworker/ecomworker.sh.j2 [hunk 4]
```diff
@@ -17,4 +17,4 @@ export NEW_RELIC_LICENSE_KEY='{{ NEWRELIC_LICENSE_KEY }}'
 
 source {{ ecommerce_worker_home }}/{{ ecommerce_worker_service_name }}_env
 # We exec so that celery is the child of supervisor and can be managed properly
-exec {{ executable }} -A ecommerce_worker worker --app ecommerce_worker.celery_app:app --concurrency={{ ECOMMERCE_WORKER_CONCURRENCY }} --loglevel=info --hostname=ecomworker.%%h --queue=ecommerce.fulfillment,ecommerce.email_marketing
+exec {{ executable }} --app ecommerce_worker.celery_app:app worker --concurrency={{ ECOMMERCE_WORKER_CONCURRENCY }} --loglevel=info --hostname=ecomworker.%%h --queue=ecommerce.fulfillment,ecommerce.email_marketing
```
Comment (pomegranited) [inline playbooks/roles/ecomworker/templates/edx/app/ecomworker/ecomworker.sh.j2]: The arguments here just look shuffled around -- is this change really necessary?

**Review by pomegranited — CHANGES_REQUESTED** [review]
@gabor-boros  This change seems to do more than just make `SINGLE_BEAT_IDENTIFIER` configurable.. Have asked for clarity, and suggested some changes.

File: playbooks/roles/edx_django_service/templates/edx/app/supervisor/conf.d.available/app-workers.conf.j2 (hunk 2 above)
Comment (gabor-boros) [reply playbooks/roles/edx_django_service/templates/edx/app/supervisor/conf.d.available/app-workers.conf.j2:13]: Remembering back to the reasoning, the argument ordering change is intentional as Celery changed the way of parsing worker/app arguments. Some stuff went to global scope of the CLI and some remained in subcommand scope. Unfortunately, changing the order to reduce diff would break CLI invocation.

File: playbooks/roles/edx_django_service/templates/edx/app/supervisor/conf.d.available/app-workers.conf.j2 (hunk 3 above)
Comment (gabor-boros) [reply playbooks/roles/edx_django_service/templates/edx/app/supervisor/conf.d.available/app-workers.conf.j2:37]: For me too (https://github.com/edx/configuration/pull/6586#discussion_r754919920)

File: playbooks/roles/ecomworker/templates/edx/app/ecomworker/ecomworker.sh.j2 (hunk 4 above)
Comment (gabor-boros) [reply playbooks/roles/ecomworker/templates/edx/app/ecomworker/ecomworker.sh.j2]: Same here: https://github.com/edx/configuration/pull/6586#discussion_r754919920

File: playbooks/roles/edxapp/defaults/main.yml (hunk 1 above)
Comment (gabor-boros) [reply playbooks/roles/edxapp/defaults/main.yml:189]: A little bit of background story that I should have provided in the description (that I'll fill out properly after this comment):

Some months ago edX platform was not capable of doing periodic scheduling of celery tasks. Therefore we added celerybeat to the platform. The only issue with celerybeat is that it cannot run more than once. If multiple celerybeat schedulers are running at the same time, the same task can be scheduled multiple times resulting in unwanted behaviours. Here comes `single-beat` into picture that ensures (using redis) that no more than one celerybeat processes are running.

> Can you explain what it is in akachanov's fork that's critical to us here?

Akachanov [removed](https://github.com/ybrs/single-beat/compare/master...akachanov:master#diff-fe06202dfa089ced8f935d1530b76acf71876400c17783f10a5ecf8c7ac3e67aL13) Tornadis that was published in 2018 for the last time and broke with the newest Celery. Using Akachanov's fork ensures that single-beat is still working with the newest Celery and he also seems to fixed some reconnect issues -- I assume hence the celerybeat process restarts, since I cannot see any errors and periodically scheduled tasks are processed as expected.

The instance has periodic reports configured to run every 10 minutes. To check the periodic report's results, please navigate to [instructor reports](https://pr29046.sandbox.opencraft.hosting/admin/instructor_task/instructortask/) and check the list of tasks. (You may see older tasks with 1 minute differences -- that was me testing the integration)

> Are the changes you've made to the single-beat invocation because of the changes in akachanov's fork, or are they due to the celery upgrade?

I had to do the changes because of Celery. akachanov's fork did not result in any code change for us. He mainly fixed internal logic in the package.

> Basically I want to know, can we retain single-beat==0.4.2 and still work with single-tenant redis installs like edX's?

EdX is not using celerybeat as far as I know. Based on the small history above, it will have no effect on edX and this will continue working for us too (as the instructor reports page shows).

> Also, the uptime for the celerybeat_scheduler on the sandbox isn't good.. is something wrong here?

I set periodic reports for instructor reports, so we can check that page on the admin site to ensure it is working properly.

Comment (pomegranited) [reply playbooks/roles/edxapp/defaults/main.yml:189]: Thank you for the explanation and expanded testing instructions @gabor-boros !

> I assume hence the celerybeat process restarts, since I cannot see any errors and periodically scheduled tasks are processed as expected.

Unfortunately with celery, "I can't see any errors" is no guarantee that things are working correctly - it has a habit of consuming and hiding task errors without logging anything. :(

But since those periodic tasks are indeed being run, then it looks like this is just business as usual for the single-beat process when run in "supervised" mode. I read a bit more about these run modes on the [single-beat README](https://github.com/ybrs/single-beat#configuration), and that seems consistent.

**Review by pomegranited — APPROVED** [review]
:+1:

- [x] I tested this on the PR sandbox.
- [x] I read through the code
- [x] <strike>I checked for accessibility issues</strike> N/A
- [x] <strike>Includes documentation</strike> N/A -- as per the [PR template checklist](https://github.com/edx/configuration/blob/master/.github/PULL_REQUEST_TEMPLATE.md), this PR does not add any defaults which need to be overridden, so no CHANGELOG entry is necessary.
- [x] <strike>I made sure any change in configuration variables is reflected in the corresponding
    client's `configuration-secure` repository.</strike> N/A

@jmbowman Can you confirm that [edX is not using celerybeat](https://github.com/edx/configuration/pull/6586#discussion_r754934716) and so this change should not affect you guys?

I'll merge this as core commiter in 2 days if there are no objections on your side. CC @natabene

File: playbooks/roles/ecomworker/templates/edx/app/ecomworker/ecomworker.sh.j2 (hunk 4 above)
Comment (Agrendalath) [inline playbooks/roles/ecomworker/templates/edx/app/ecomworker/ecomworker.sh.j2]: @gabor-boros, don't we need to define `-A ecommerce_worker` anymore? I see that we've retained it for `edx_django_service_name` workers.

File: playbooks/roles/edx_django_service/templates/edx/app/supervisor/conf.d.available/app-workers.conf.j2 [hunk 5]
```diff
@@ -10,7 +10,7 @@ directory={{ edx_django_service_code_dir }}
...
 
-command={{ edx_django_service_home }}/{{ edx_django_service_name }}-workers.sh worker -A {{ edx_django_service_name }} --app {{ edx_django_service_name }}.celery:app --loglevel=info --queue={{ w.queue }} --hostname={{ edx_django_service_name }}.{{ w.queue }}.%%h --concurrency=1 {{ '--maxtasksperchild ' + w.max_tasks_per_child|string if w.max_tasks_per_child is defined else '' }} {{ '--without-heartbeat' if not edx_django_service_celery_heartbeat_enabled|bool else '' }}
+command={{ edx_django_service_home }}/{{ edx_django_service_name }}-workers.sh --app {{ edx_django_service_name }}.celery:app worker -A {{ edx_django_service_name }} --loglevel=info --queue={{ w.queue }} --hostname={{ edx_django_service_name }}.{{ w.queue }}.%%h --concurrency=1 {{ '--max-tasks-per-child ' + w.max_tasks_per_child|string if w.max_tasks_per_child is defined else '' }} {{ '--without-heartbeat' if not edx_django_service_celery_heartbeat_enabled|bool else '' }}
 killasgroup=true
 stopwaitsecs={{ w.stopwaitsecs | default(edx_django_service_default_stopwaitsecs) }}
 ; Set autorestart to `true`. The default value for autorestart is `unexpected`, but celery < 4.x will exit
```
Comment (Agrendalath) [inline playbooks/roles/edx_django_service/templates/edx/app/supervisor/conf.d.available/app-workers.conf.j2]: @gabor-boros, as we're upgrading Celery, it might be worth removing these parts here and in the supervisor. We can also ignore it - I'll leave this up to you.

File: playbooks/roles/edxapp/defaults/main.yml (hunk 1 above)
Comment (Agrendalath) [reply playbooks/roles/edxapp/defaults/main.yml:189]: @gabor-boros, can we please update the CHANGELOG?

**Review by Agrendalath — APPROVED** [review]
:+1:

* [x]  I tested this: I've checked heartbeats and periodic tasks in the sandbox
* [x]  I read through the code
* [x]  I checked for accessibility issues: n/a
* [ ]  Includes documentation: no :warning: 
* [x] I made sure any change in configuration variables is reflected in the corresponding client's `configuration-secure` repository: n/a

File: playbooks/roles/ecomworker/templates/edx/app/ecomworker/ecomworker.sh.j2 (hunk 4 above)
Comment (pomegranited) [reply playbooks/roles/ecomworker/templates/edx/app/ecomworker/ecomworker.sh.j2]: Good catch @Agrendalath  -- this seems important.

@gabor-boros  could you fix this before I merge?

**Comment by gabor-boros** [comment]
@Agrendalath I believe all the requested changes are addressed. The app server is provisioning with the new changes.

**Comment by openedx-webhooks** [comment]
@gabor-boros 🎉 Your pull request was merged!  Please take a moment  to [answer a two question survey](https://docs.google.com/forms/d/e/1FAIpQLSceJOyGJ6JOzfy6lyR3T7EW_71OWUnNQXp68Fymsk3MkNoSDg/viewform?usp=pp_url&entry.1671973413=edx/configuration&entry.867055334=https://github.com/edx/configuration/pull/6586&entry.1484655318=https://github.com/gabor-boros&entry.752974735=2021-10-18+05:43&entry.1917517419=2021-12-07+00:31&entry.2133058324=Yes) so we can improve your experience in the future.
